Implement parsing for array(subquery) expressions

smakalias opened the pull request:
## Summary of the changes / Why this improves CrateDB

Update parser grammar and add a new AST node: `ArraySubQueryExpression` for allowing parsing of the array(subquery) feature.

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed

smakalias commented:
I would appreciate an opinion on my choice to add this as an expression as I believe the array(subquery) is an array constrictor basically and should be able to be used anywhere, not just in the projection list. See e.g. in my test `testArrayConstructorSubSelectBuilder` where I used it for an array comparison expression.

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/tree/ArraySubQueryExpression.java:
```diff
@@ -0,0 +1,59 @@
+    @Nullable
+    private final Expression subqueryExpression;
+
+    public ArraySubQueryExpression(@Nullable Expression subqueryExpression) {
```
I think it should be possible to remove the `Nullable` annotation?

mfussenegger replied:
Would it make sense to be more restrictive here and only allow `SubqueryExpression` instances? Given that the name of this class `ArraySubQueryExpression` implies that only subqueries are allowed.

smakalias replied:
good point!

smakalias replied:
Thing is that if I use the specific type in the constructor, then I need to cast the visit return type to that specific type in the ASTBuilder: 
```
    @Override
    public Node visitArraySubquery(SqlBaseParser.ArraySubqueryContext ctx) {
        return new ArraySubQueryExpression((SubqueryExpression) visit(ctx.subqueryExpression()));
    }
```
where in all other case the more general class Expression is being used.

What do you think is better? I would go for the generic type Expression

mfussenegger replied:
Although consistency is nice, I think it's more important to catch errors early and make it impossible to construct invalid objects. .E.g. creating a `ArraySubQueryExpression` with a literal inside doesn't make any sense. And it would move the validation/error handling to all consumers of the `ArraySubQueryExpression`

So I'd be in favor of casting to `SubqueryExpression`

smakalias commented on sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java:
```diff
@@ -1229,6 +1237,14 @@ public void testSubSelects() throws Exception {
         printStatement("select * from (select * from (select * from foo) f1) f2");
     }
 
+//    @Test
```
note to self: removed those

smakalias commented:
@mfussenegger done all the suggested changes and added one more test. 
I also removed the WIP since it will be merged to a feature branch and continue from there.
Please check again. Thanks

mfussenegger commented on sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java:
```diff
@@ -749,6 +749,21 @@ public void testArrayConstructorStmtBuilder() {
+
+    @Test
+    public void testArrayConstructorSubSelectBuilderNoSubQueryThrowsParsingException() throws Exception {
+        expectedException.expect(ParsingException.class);
```
afaik this will only test the first statement and then exit, not verifying the second case

smakalias commented:
squashed commits
